### PR TITLE
add reverse tokenizer

### PR DESCRIPTION
**Pull request by aquatiko**
A simple method for a reversible tokenizer. 

Closes: https://github.com/JuliaText/WordTokenizers.jl/issues/12
Closes: https://github.com/JuliaText/TextAnalysis.jl/issues/107

- [X] Add function
- [X] Add documentation
- [X] Add tests

### CODE REVIEW

File: src/reversible_tokenize.jl [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    for ind in eachindex(instring)
+        c = instring[ind]
+        c_p = ind > 1 ? instring[ind-1] : c
+        c_n = ind < length(instring) ? instring[ind+1] : c
+        
+        if .!(is_weird(c))
+            ans = ans*c
+        else
+            if .!(isspace(c_p))
+                ans = ans * " " * MERGESYMBOL
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
                ans *= " " * MERGESYMBOL
```

File: src/reversible_tokenize.jl [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+function rev_tokenize(instring::AbstractString)
+    ans = ""
+    for ind in eachindex(instring)
+        c = instring[ind]
+        c_p = ind > 1 ? instring[ind-1] : c
+        c_n = ind < length(instring) ? instring[ind+1] : c
+        
+        if .!(is_weird(c))
+            ans = ans*c
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
            ans *= c
```

File: src/reversible_tokenize.jl [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# A simple reversible tokenizer
+
+MERGESYMBOL = '~'
+
+function is_weird(c::AbstractChar)
+    return .!(isletter(c) ||isnumeric(c) || isspace(c))
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
    return !(isletter(c) ||isnumeric(c) || isspace(c))
```

the `.!` is broadcasting and this isn't breadcasting

File: src/reversible_tokenize.jl [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+end
+
+
+function rev_tokenize(instring::AbstractString)
+    ans = ""
+    for ind in eachindex(instring)
```
Comment (oxinabox) [inline src/reversible_tokenize.jl:43]: Make sure to include a test with multibyte unicode characters

File: src/reversible_tokenize.jl [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        c = instring[ind]
+        c_p = ind > 1 ? instring[ind-1] : c
+        c_n = ind < length(instring) ? instring[ind+1] : c
+        
+        if .!(is_weird(c))
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
        if !is_weird(c)
```

File: src/reversible_tokenize.jl [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            ans = ans*c
+        else
+            if .!(isspace(c_p))
+                ans = ans * " " * MERGESYMBOL
+            end
+            ans = ans * c
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
            ans *=  c
```

**Comment by aquatiko** [comment]
One error that I was facing is initialisation of `MERGESYMBOL` with some unicode character (say `\leftrightarrow`), for now I have initialised it with `'~'`.  But this might cause some unexpected result if the input string contains it. Still thinking how to fix it.

File: src/reversible_tokenize.jl [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            if .!(isspace(c_n) && is_weird(c_n))
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
            if !isspace(c_n) && is_weird(c_n)
```

File: src/reversible_tokenize.jl [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+                ans = ans * MERGESYMBOL * " "
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
                ans *= MERGESYMBOL * " "
```

File: src/reversible_tokenize.jl [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        c = instring[ind]
+        c_p = ind > 1 ? instring[ind-1] : c
+        c_n = ind < length(instring) ? instring[ind+1] : c
+        c_pp = ind > 2 ? instring[ind-2] : c
+        c_nn = ind < length(instring) - 1 ? instring[ind+2] : c
+        
+        if c * c_n == ' ' * MERGESYMBOL && is_weird(c_nn)
+            ind += 2
+        elseif is_weird(c) && c_n * c_nn == MERGESYMBOL * ' '
+            ans = ans * c
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
            ans *=  c
```

File: src/reversible_tokenize.jl [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            ind += 3
+        else
+            ans = ans * c
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
            ans *= c
```

File: src/reversible_tokenize.jl [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            ind += 1
+        end
+    end
+    ans
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
    return ans
```

File: src/reversible_tokenize.jl [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+            if .!(isspace(c_n) && is_weird(c_n))
+                ans = ans * MERGESYMBOL * " "
+            end
+        end
+    end
+    ans
+end
+
+
+function rev_detokenize(instring::AbstractString)
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]: I feel like this should take a array of strings

File: src/reversible_tokenize.jl [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            if .!(isspace(c_p))
+                ans = ans * " " * MERGESYMBOL
+            end
+            ans = ans * c
+            if .!(isspace(c_n) && is_weird(c_n))
+                ans = ans * MERGESYMBOL * " "
+            end
+        end
+    end
+    ans
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
    return ans
```

Comment (oxinabox) [reply src/reversible_tokenize.jl]: A tokenizer should return a collection of tokens.

**Review by oxinabox — COMMENTED** [review]
thanks,
I have not done a proper review just a quick pass-over.

One thought worth considering is if the could be done so that it wraps *any* tokenizer,
and makes it reversible.

So one could say call `reveriable(penn_tokenizer)` and it would return a tokenizer function that is reverable but uses the penn tokenizer to decide what tokens to make.
I am not sure how feasible that is, it would require some thought.

File: src/reversible_tokenize.jl (hunk 12 above)
Comment (aquatiko) [reply src/reversible_tokenize.jl]: @oxinabox Actually, the detokenization here is achieved by leaving markers in the original string and the thus the method takes input as a single string in a tokenized representation with markers.

**Comment by oxinabox** [comment]
>One error that I was facing is initialisation of `MERGESYMBOL` with some unicode character (say `\leftrightarrow`), for now I have initialised it with `'~'`. But this might cause some unexpected result if the input string contains it. Still thinking how to fix it.

You use a unicode character that can't appear in text. Like something from the private use area
e.g.
https://www.fileformat.info/info/unicode/char/e302/index.htm

**Comment by oxinabox** [comment]
You'll need to rebase this should be a simple rebase.
Also how do you feel about restructuring this to use the fast new tokenbuffer that was just merged.
https://github.com/JuliaText/WordTokenizers.jl/blob/master/src/words/fast.jl

File: src/reversible_tokenize.jl (hunk 12 above)
Comment (oxinabox) [reply src/reversible_tokenize.jl]: Yes, that is also a problem with the tokenzier.
A tokenizer must return an iterable collection of tokens.
That is how we define it.
Just adding `split` to your answer will do this though.

**Comment by MikeInnes** [comment]
I don't think this will gain anything from the more complex tokenisation infra, given how simple it is. But you definitely should avoid `*` for string concatenation. Instead have a look at `IOBuffer`.

**Comment by aquatiko** [comment]
@oxinabox I have pushed the asked changes :)

**Comment by oxinabox** [comment]
Cool. the `IOBuffer` is both clearer and faster.
I will review more closely once tests and docs are in.


A note on format (which I just noticed  now).
Prefered spacing is:

 - `foo(a, b)`, not `foo(a,b)`, always space after a comma in a function call
 - `1 + 1`, not `1+1`,  basically space around all operators
 - exception to the above rule is the kwargs should be `foo(a, b; bar=1)` not `foo(a, b; bar = 1`)

(at some point I will have a style guide linked in the repo.)

File: src/reversible_tokenize.jl [hunk 14]
```diff
@@ -1,5 +1,23 @@
...
+using the rev_detokenize.
+It uses a character based approach for splitting and re-merging.
+
+Parameters:
+
+- instring		= Input string to be tokenized 
+- token			= Collection to tokens i.e String Array
+
+"""
 MERGESYMBOL = '~'
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]: Use a character from unicode private use area. e.g. `\ue302`

Comment (aquatiko) [reply src/reversible_tokenize.jl]: I was trying to do that for a while... But this error persists if I use any of the `\` symbols for `MERGESYMBOL` , thus I added `~`
```
Some of 1 ↹,↹ 0000 households (↹ usually ↹, a minority ↹) ate breakfast ↹.
StringIndexError("Some of 1 ↹,↹ 0000 households (↹ usually ↹, a minority ↹) ate breakfast ↹.", 12)

Stacktrace:
 [1] string_index_err(::String, ::Int64) at ./strings/string.jl:12
 [2] getindex_continued(::String, ::Int64, ::UInt32) at ./strings/string.jl:218
 [3] rev_detokenize(::Array{SubString{String},1}) at ./strings/string.jl:211
 [4] top-level scope at In[21]:1

```
There wasn't any trouble while using `rev_tokenize`, just when I use `rev_detokenize` like this way.

File: src/WordTokenizers.jl [hunk 15]
```diff
@@ -1,20 +1,21 @@
...
 export poormans_tokenize, punctuation_space_tokenize,
        penn_tokenize, improved_penn_tokenize, nltk_word_tokenize,
        tokenize,
        rulebased_split_sentences,
        split_sentences,
-       set_tokenizer, set_sentence_splitter
+       set_tokenizer, set_sentence_splitter,
+       rev_tokenize, rev_detokenize
 
-include("words/fast.jl")
```
Comment (oxinabox) [inline src/WordTokenizers.jl:10]: I am guessing this was deleted by mistake during a rebase.
Removing it is why tests are now failing

**Comment by oxinabox** [comment]
Can we have more tests?

In particular some with:
 - Multiple consecutive spaces
 - multibyte unicode characters

I suspect multibyte unicode characters may be a problem as 
not all positions in a string can be indexed to

Take a read of https://docs.julialang.org/en/v1/manual/strings/#Unicode-and-UTF-8-1
and https://docs.julialang.org/en/v1/base/strings/#Base.thisind

**Comment by aquatiko** [comment]
@oxinabox Incorporated asked additions :)

File: src/reversible_tokenize.jl [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+using the rev_detokenize.
+It uses a character based approach for splitting and re-merging.
+
+Parameters:
+
+- instring		= Input string to be tokenized 
+- token			= Collection to tokens i.e String Array
+
+"""
+MERGESYMBOL = '⇆'
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]: I still think the symbol should be from the private use area.
(i.e. characters that on most font's will not even display and that will not display consistantly between fonts.

Also this line needs to be declared `const`

Comment (oxinabox) [reply src/reversible_tokenize.jl]: https://en.wikipedia.org/wiki/Private_Use_Areas

File: test/reversible_tok.jl [hunk 17]
```diff
@@ -0,0 +1,26 @@
+using Test
+using WordTokenizers
+
```
Comment (oxinabox) [inline test/reversible_tok.jl:3]: use named `@testset`s rather than comments to make it clear what these tests each do

(You can nest them, even)

File: src/reversible_tokenize.jl [hunk 18]
```diff
@@ -49,25 +49,26 @@ function rev_tokenizer(instring::AbstractString)
...
+        c_pp = thisind(instring, ind) > nextind(instring, 1) ? instring[prevind(instring, prevind(instring, ind))] : c
+        c_nn = thisind(instring, ind) < prevind(instring, last_index) ? instring[nextind(instring, nextind(instring, ind))] : c
         
         if c * c_n == ' ' * MERGESYMBOL && is_weird(c_nn)
-            ind += 2
+            ind = nextind(instring, nextind(instring, ind))
         elseif is_weird(c) && c_n * c_nn == MERGESYMBOL * ' '
             write(ans, c)
-            ind += 3
+            ind = nextind(instring, nextind(instring, nextind(instring, ind)))
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]: Do you think it would add clarity to define a helper function,

`nth_ind(instring, start_ind, n)` that recursively calls `nextind`, `n` times ?

Or perhaps just a comment saying "Advance 3 characters", "advance 2 characters" ?

Or perhaps this is fine

**Comment by oxinabox** [comment]
Looking pretty good,

I think it just needs these few minor things and adding to the readme.

File: src/reversible_tokenize.jl (hunk 18 above)
Comment (aquatiko) [reply src/reversible_tokenize.jl]: A helper function was something I was looking for while writing this,
I will add the function for the same.

File: README.md [hunk 19]
```diff
@@ -61,6 +61,9 @@ The word tokenizers basically assume sentence splitting has already been done.
  - **Improved Penn Tokenizer:** (`improved_penn_tokenize`) NLTK's improved Penn Treebank Tokenizer. Very similar to the original, some improvements on punctuation and contractions. This matches to NLTK's `nltk.tokenize.TreeBankWordTokenizer.tokenize`
  - **NLTK Word tokenizer:** (`nltk_word_tokenize`) NLTK's even more improved version of the Penn Tokenizer. This version has better unicode handling and some other changes. This matches to the most commonly used `nltk.word_tokenize`, minus the sentence tokenizing step. **(default tokenizer)**
```
Comment (oxinabox) [inline README.md]:
```suggestion
```

File: README.md [hunk 20]
```diff
@@ -61,6 +61,9 @@ The word tokenizers basically assume sentence splitting has already been done.
...
 
+- **Reversible Tokenizer:** (`rev_tokenizer` and `rev_detokenizer`) This tokenizer splits on punctuations, space amd special symbols. The generated tokens can be de-tokenized by using the `rev_detokenizer` function into the state before tokenization.
```
Comment (oxinabox) [inline README.md]:
```suggestion
- **Reversible Tokenizer:** (`rev_tokenizer` and `rev_detokenizer`) This tokenizer splits on punctuations, space and special symbols. The generated tokens can be de-tokenized by using the `rev_detokenizer` function into the state before tokenization. 
```

File: src/reversible_tokenize.jl [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+```
+
+The rev_tokenize tokenizer splits into token based on space, punctuations and special symbols and in 
+addition it leaves some MERGESYMBOLS for the tokens to be re-arranged when needed 
+using the rev_detokenize.
+It uses a character based approach for splitting and re-merging.
+
+Parameters:
+
+- instring		= Input string to be tokenized
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
- instring: Input string to be tokenized 
```

File: src/reversible_tokenize.jl [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+"""
+A simple reversible tokenizer
+
+```
+tokenized = rev_tokenizer(instring)
+de_tokenized = rev_detokenizer(token)
+```
+
+The rev_tokenize tokenizer splits into token based on space, punctuations and special symbols and in 
+addition it leaves some MERGESYMBOLS for the tokens to be re-arranged when needed
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
addition it leaves some merge-symbols (`'\ue302'`) for the tokens to be re-arranged when needed 
```

File: src/reversible_tokenize.jl [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+
+The rev_tokenize tokenizer splits into token based on space, punctuations and special symbols and in 
+addition it leaves some MERGESYMBOLS for the tokens to be re-arranged when needed 
+using the rev_detokenize.
+It uses a character based approach for splitting and re-merging.
+
+Parameters:
+
+- instring		= Input string to be tokenized 
+- token			= Collection to tokens i.e String Array
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
- token: Collection to tokens i.e String Array
```

File: src/reversible_tokenize.jl [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+
+"""
+const global MERGESYMBOL = '\ue302'
```
Comment (oxinabox) [inline src/reversible_tokenize.jl]:
```suggestion
const MERGESYMBOL = '\ue302'
```

`global` does nothing when called at global scope

Comment (aquatiko) [reply src/reversible_tokenize.jl]: -Updated

**Comment by oxinabox** [comment]
LGTM, thanks!

**Comment by jekbradbury** [comment]
Shouldn’t the function name be `...tokenize` like the others, not `...tokenizer`?

**Comment by oxinabox** [comment]
Bother, missed that. Yes
